File: mobile/native/android/apps/ensu/app/src/main/java/io/ente/ensu/llm/ModelDownloader.kt
package io.ente.ensu.llm

import android.content.Context
import android.os.Environment
import android.util.Log
import io.ente.ensu.bindings.CancellationToken
import io.ente.ensu.bindings.ModelDownloadCallback
import io.ente.ensu.bindings.ModelDownloadCore
import io.ente.ensu.bindings.ModelDownloadProgress
import io.ente.ensu.bindings.LlmException
import io.ente.ensu.bindings.ModelTarget
import io.ente.ensu.bindings.LegacyModels
import io.ente.ensu.bindings.migrateMobileModels
import io.ente.ensu.bindings.transcriptionModelTarget
import io.ente.ensu.bindings.voiceActivityModelTarget
import io.ente.ensu.bindings.uniffiEnsureInitialized
import java.io.File
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.withContext

class ModelDownloader(context: Context) {
    private val appContext = context.applicationContext
    private val modelsDir = File(appContext.noBackupFilesDir, "models")
    private val legacyDir = appContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
        ?.let { File(it, "llm") }
    private val legacyTranscriptionDir = File(appContext.dataDir, "app_ensu_transcription_models")
    private val downloadMutex = Mutex()
    private val core: ModelDownloadCore
    val transcriptionTarget: ModelTarget
    val voiceActivityTarget: ModelTarget

    init {
        uniffiEnsureInitialized()
        ModelDownloadJobService.attach(appContext)
        core = ModelDownloadCore(modelsDir.absolutePath)
        transcriptionTarget = transcriptionModelTarget()
        voiceActivityTarget = voiceActivityModelTarget()
    }

    fun needsMigration(): Boolean =
        legacyDir?.exists() == true || legacyTranscriptionDir.exists()

    fun migrate(legacyModelUrl: String?, legacyMmprojUrl: String?): String? {
        File(appContext.filesDir, "llm").deleteRecursively()
        return migrateMobileModels(
            modelsDir.absolutePath,
            LegacyModels(
                legacyDir?.absolutePath,
                legacyTranscriptionDir.absolutePath,
                legacyModelUrl,
                legacyMmprojUrl
            )
        )
    }

    val isDownloadActive: Boolean get() = core.isDownloadActive()

    fun modelDir(target: ModelTarget): File = File(core.modelDir(target))

    fun llmModelPath(target: ModelTarget): File? =
        core.llmModelPath(target)?.let(::File)

    fun llmMmprojPath(target: ModelTarget): File? =
        core.llmMmprojPath(target)?.let(::File)

    fun voiceActivityModelPath(): File = File(core.voiceActivityModelPath())

    fun isDownloaded(target: ModelTarget): Boolean = core.isDownloaded(target)

    fun removeDownloaded(target: ModelTarget): Boolean = core.removeDownloaded(target)

    suspend fun estimateDownloadSize(target: ModelTarget): Long? = withContext(Dispatchers.IO) {
        core.estimatedDownloadSize(target)
    }

    suspend fun download(
        targets: List<ModelTarget>,
        onProgress: (DownloadProgress) -> Unit
    ): Unit = withContext(Dispatchers.IO) {
        if (!downloadMutex.tryLock()) throw LlmException.Cancelled()
        try {
            downloadLocked(targets, onProgress)
        } finally {
            downloadMutex.unlock()
        }
    }

    private suspend fun downloadLocked(
        targets: List<ModelTarget>,
        onProgress: (DownloadProgress) -> Unit
    ) {
        if (targets.all { core.isDownloaded(it) }) return

        val token = CancellationToken()
        ModelDownloadJobService.begin { token.cancel() }
        try {
            coroutineScope {
                val download = async {
                    core.download(
                        targets,
                        object : ModelDownloadCallback {
                            override fun onProgress(progress: ModelDownloadProgress) {
                                progress.logLine?.let { Log.i("ModelDownloader", it) }
                                ModelDownloadJobService.update(progress.percent, progress.totalBytes == null)
                                onProgress(DownloadProgress(progress.percent, progress.status))
                            }
                        },
                        token
                    )
                }
                try {
                    download.await()
                } catch (e: CancellationException) {
                    token.cancel()
                    throw e
                }
            }
        } finally {
            ModelDownloadJobService.end()
        }
    }
}

File: mobile/native/apple/apps/ensu/Ensu/Services/ModelDownloader.swift
import BackgroundTasks
import Foundation

private let logger = EnsuLogging.shared.logger("ModelDownloader")

final class ModelDownloader {
    private let core: ModelDownloadCore
    private let activeLock = NSLock()
    private var activeToken: CancellationToken?
    let transcriptionTarget: ModelTarget
    let voiceActivityTarget: ModelTarget

    @MainActor
    init() {
        let baseDir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        var modelsDir = baseDir.appendingPathComponent("models", isDirectory: true)
        try? FileManager.default.createDirectory(at: modelsDir, withIntermediateDirectories: true)
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try? modelsDir.setResourceValues(values)
        core = ModelDownloadCore(modelsDir: modelsDir.path)
        transcriptionTarget = transcriptionModelTarget()
        voiceActivityTarget = voiceActivityModelTarget()
        let settings = UserDefaults.standard
        let pendingSelection = settings.object(forKey: "ensu.model.id") == nil
        let legacyModelUrl = pendingSelection && settings.bool(forKey: "ensu.model.use_custom")
            ? settings.string(forKey: "ensu.model.url")
            : nil
        let presetId = migrateMobileModels(
            modelsDir: modelsDir.path,
            legacy: LegacyModels(
                llmDir: baseDir.appendingPathComponent("llm", isDirectory: true).path,
                transcriptionDir: baseDir.appendingPathComponent("transcription", isDirectory: true).path,
                modelUrl: legacyModelUrl,
                mmprojUrl: settings.string(forKey: "ensu.model.mmproj")
            )
        )
        if pendingSelection {
            settings.set(presetId ?? "", forKey: "ensu.model.id")
        }
        settings.removeObject(forKey: "ensu.model.use_custom")
        settings.removeObject(forKey: "ensu.model.url")
        settings.removeObject(forKey: "ensu.model.mmproj")
    }

    static func registerBackgroundTask() {
        if #available(iOS 26.0, *) {
            ModelDownloadBackgroundTask.register()
        }
    }

    func modelDir(_ target: ModelTarget) -> URL {
        URL(fileURLWithPath: core.modelDir(target: target))
    }

    func llmModelPath(_ target: ModelTarget) -> URL? {
        core.llmModelPath(target: target).map { URL(fileURLWithPath: $0) }
    }

    func llmMmprojPath(_ target: ModelTarget) -> URL? {
        core.llmMmprojPath(target: target).map { URL(fileURLWithPath: $0) }
    }

    func voiceActivityModelPath() -> URL {
        URL(fileURLWithPath: core.voiceActivityModelPath())
    }

    func isDownloaded(_ target: ModelTarget) -> Bool {
        core.isDownloaded(target: target)
    }

    func removeDownloaded(_ target: ModelTarget) -> Bool {
        core.removeDownloaded(target: target)
    }

    func cancel() {
        activeLock.withLock { activeToken }?.cancel()
    }

    func estimateDownloadSize(_ target: ModelTarget) async -> Int64? {
        await Task.detached(priority: .utility) { [core] in
            core.estimatedDownloadSize(target: target)
        }.value
    }

    func download(
        targets: [ModelTarget],
        onProgress: @escaping (DownloadProgress) -> Void
    ) async throws {
        let token = CancellationToken()
        let started = activeLock.withLock { () -> Bool in
            if activeToken != nil { return false }
            activeToken = token
            return true
        }
        guard started else { throw DownloadAlreadyActiveError() }
        defer { activeLock.withLock { activeToken = nil } }

        if targets.allSatisfy({ self.core.isDownloaded(target: $0) }) {
            return
        }

        onProgress(DownloadProgress(percent: 0, status: "Starting download..."))

        if #available(iOS 26.0, *) {
            ModelDownloadBackgroundTask.begin {
                token.cancel()
            }
        }
        var succeeded = false
        defer {
            if #available(iOS 26.0, *) {
                ModelDownloadBackgroundTask.end(success: succeeded)
            }
        }

        let core = core
        let downloadTask = Task.detached(priority: .utility) {
            try Task.checkCancellation()
            let callback = ModelDownloadCallbackSink { progress in
                if let line = progress.logLine {
                    logger.info(line)
                }
                if #available(iOS 26.0, *) {
                    ModelDownloadBackgroundTask.update(
                        downloadedBytes: progress.downloadedBytes,
                        totalBytes: progress.totalBytes
                    )
                }
                onProgress(DownloadProgress(percent: Int(progress.percent), status: progress.status))
            }
            try core.download(targets: targets, callback: callback, cancellation: token)
        }

        try await withTaskCancellationHandler {
            try await downloadTask.value
        } onCancel: {
            downloadTask.cancel()
            token.cancel()
        }
        succeeded = true
    }
}

private struct DownloadAlreadyActiveError: Error {}

private final class ModelDownloadCallbackSink: ModelDownloadCallback, @unchecked Sendable {
    private let onProgressHandler: (ModelDownloadProgress) -> Void

    init(onProgress: @escaping (ModelDownloadProgress) -> Void) {
        self.onProgressHandler = onProgress
    }

    func onProgress(progress: ModelDownloadProgress) {
        onProgressHandler(progress)
    }
}

@available(iOS 26.0, *)
private enum ModelDownloadBackgroundTask {
    private static let identifier = "io.ente.ensu.model-download"
    private static let lock = NSLock()
    private static var task: BGContinuedProcessingTask?
    private static var onExpiration: (() -> Void)?
    private static var downloadActive = false

    static func register() {
        BGTaskScheduler.shared.register(forTaskWithIdentifier: identifier, using: nil) { bgTask in
            guard let bgTask = bgTask as? BGContinuedProcessingTask else {
                bgTask.setTaskCompleted(success: false)
                return
            }
            adopt(bgTask)
        }
    }

    static func begin(onExpiration: @escaping () -> Void) {
        lock.lock()
        downloadActive = true
        Self.onExpiration = onExpiration
        lock.unlock()

        let request = BGContinuedProcessingTaskRequest(
            identifier: identifier,
            title: "Downloading model",
            subtitle: ""
        )
        request.strategy = .fail
        do {
            try BGTaskScheduler.shared.submit(request)
        } catch {
            logger.warning("Background download task not scheduled", details: "\(error)")
        }
    }

    static func update(downloadedBytes: Int64, totalBytes: Int64?) {
        lock.lock()
        let task = task
        lock.unlock()
        guard let task, let totalBytes, totalBytes > 0 else { return }
        task.progress.totalUnitCount = totalBytes
        task.progress.completedUnitCount = min(downloadedBytes, totalBytes)
    }

    static func end(success: Bool) {
        lock.lock()
        downloadActive = false
        onExpiration = nil
        let task = task
        Self.task = nil
        lock.unlock()
        task?.setTaskCompleted(success: success)
    }

    private static func adopt(_ bgTask: BGContinuedProcessingTask) {
        lock.lock()
        guard downloadActive else {
            lock.unlock()
            bgTask.setTaskCompleted(success: true)
            return
        }
        task = bgTask
        lock.unlock()

        bgTask.expirationHandler = {
            lock.lock()
            let expired = task === bgTask
            let handler = onExpiration
            if expired {
                task = nil
            }
            lock.unlock()
            if expired {
                handler?()
                bgTask.setTaskCompleted(success: false)
            }
        }
    }
}
